From e3125584b68f90b9b7d1316f07705ad48d2f6c43 Mon Sep 17 00:00:00 2001
From: Dan McGee <dan@archlinux.org>
Date: Tue, 27 Mar 2012 17:22:40 -0500
Subject: [PATCH] Fixes for GCC 4.7.0

Fixes the following compile error exposed with GCC 4.7.0:

    libarchive/archive_string.c: In function 'cesu8_to_unicode':
    libarchive/archive_string.c:2450:11: error: 'wc' may be used uninitialized in this function [-Werror=uninitialized]
    cc1: all warnings being treated as errors

As well as a test failure that depends on signed integer wraparound,
which is a very bad thing to do in C [1]. Calculate the smallest
possible integer for the type, regardless of whether time_t is 4 or 8 or
27 bytes wide, as long as it is in two's complement representation. We
can then compare it directly to the value archive_entry_mtime()
returned.

[1] http://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Signed-Overflow-Examples.html
---
 libarchive/archive_string.c              |    3 ++-
 libarchive/test/test_read_format_mtree.c |    7 ++++---
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/libarchive/archive_string.c b/libarchive/archive_string.c
index 2728a37..2b56a48 100644
--- a/libarchive/archive_string.c
+++ b/libarchive/archive_string.c
@@ -2447,11 +2447,12 @@ combine_surrogate_pair(uint32_t uc, uint32_t uc2)
 static int
 cesu8_to_unicode(uint32_t *pwc, const char *s, size_t n)
 {
-	uint32_t wc, wc2;
+	uint32_t wc = 0;
 	int cnt;
 
 	cnt = _utf8_to_unicode(&wc, s, n);
 	if (cnt == 3 && IS_HIGH_SURROGATE_LA(wc)) {
+		uint32_t wc2 = 0;
 		if (n - 3 < 3) {
 			/* Invalid byte sequence. */
 			goto invalid_sequence;
diff --git a/libarchive/test/test_read_format_mtree.c b/libarchive/test/test_read_format_mtree.c
index 0d86bd4..d5527e0 100644
--- a/libarchive/test/test_read_format_mtree.c
+++ b/libarchive/test/test_read_format_mtree.c
@@ -136,10 +136,11 @@ test_read_format_mtree1(void)
 	/* The value in the file is MIN_INT64_T, but time_t may be narrower. */
 	/* Verify min_time is the smallest possible time_t. */
 	min_time = archive_entry_mtime(ae);
+	/* this calculates the minimum value in two's complement */
+	t = 1;
+	t <<= (sizeof(t) * 8 - 1);
 	assert(min_time <= 0);
-	/* Simply asserting min_time - 1 > 0 breaks with some compiler optimizations. */
-	t = min_time - 1;
-	assert(t > 0);
+	assert(min_time == t);
 
 	/* toooldfile is 1 sec older, which should overflow and get returned
 	 * with the same value. */
-- 
1.7.9.5

